pkg/kube/controllers/quarksstatefulset/quarksstatefulset_reconciler.go

```diff
@@ -281,36 +254,39 @@ func (r *ReconcileQuarksStatefulSet) generateSingleStatefulSet(qStatefulSet *qst
 		}
 		annotations[qstsv1a1.AnnotationZones] = string(zonesBytes)
 
-		podLabels[qstsv1a1.LabelAZIndex] = strconv.Itoa(zoneIndex)
-		podLabels[qstsv1a1.LabelAZName] = zoneName
-
-		podAnnotations[qstsv1a1.AnnotationZones] = string(zonesBytes)
-
 		statefulSet = r.updateAffinity(statefulSet, qStatefulSet.Spec.ZoneNodeLabel, zoneName)
 	}
+	labels[qstsv1a1.LabelAZIndex] = strconv.Itoa(zoneIndex)
+	labels[qstsv1a1.LabelQStsName] = qStatefulSet.GetName()
 
-	podLabels[qstsv1a1.LabelAZIndex] = strconv.Itoa(zoneIndex)
-	podLabels[qstsv1a1.LabelQStsName] = qStatefulSet.GetName()
-	podLabels[manifest.LabelDeploymentVersion] = fmt.Sprintf("%d", version)
-
-	statefulSet.Spec.Template.SetLabels(podLabels)
-	statefulSet.Spec.Template.SetAnnotations(podAnnotations)
-
-	r.injectContainerEnv(&statefulSet.Spec.Template.Spec, zoneIndex, zoneName, qStatefulSet.Spec.Template.Spec.Replicas)
-
-	annotations[qstsv1a1.AnnotationVersion] = fmt.Sprintf("%d", version)
+	annotations[qstsv1a1.AnnotationVersion] = strconv.Itoa(version)
+	annotations[statefulset.AnnotationCanaryRolloutEnabled] = "true"
 
 	// Set updated properties
-	statefulSet.SetName(fmt.Sprintf("%s-v%d", statefulSetNamePrefix, version))
-	statefulSet.SetLabels(labels)
+	statefulSet.Spec.Template.SetLabels(mergeMaps(statefulSet.Spec.Template.GetLabels(), labels))
+	statefulSet.Spec.Template.SetAnnotations(mergeMaps(statefulSet.Spec.Template.GetAnnotations(), annotations))
+	statefulSet.SetName(statefulSetNamePrefix)
+	statefulSet.SetLabels(mergeMaps(statefulSet.GetLabels(), labels))
 	statefulSet.Spec.Selector = &metav1.LabelSelector{
 		MatchLabels: labels,
 	}
-	statefulSet.SetAnnotations(annotations)
+	statefulSet.SetAnnotations(mergeMaps(statefulSet.GetAnnotations(), annotations))
 
+	r.injectContainerEnv(&statefulSet.Spec.Template.Spec, zoneIndex, zoneName, qStatefulSet.Spec.Template.Spec.Replicas)
 	return statefulSet, nil
 }
 
+func mergeMaps(maps ...map[string]string) map[string]string {
```

Comment:
Could be replaced by "deepMerge" or so

Reply:
https://godoc.org/github.com/imdario/mergo#MergeWithOverwrite

Reply:
> "for value type attributes on the dst struct"

@manno The `MergeWithOverwrite` can only be used for structs, right?

Reply:
@c0d1ngm0nk3y I don't think I ever used it myself, just saw it in the code once and the web page says:

> Mergo: merging Go structs and maps since 2013.

Reply:
Strange, I got a reflection error that it is a map value. I will have another look tomorrow.

Reply:
We decided to stay with the old implementation for the following reasons:
* It's simpler, because no reflection is needed
* No error handling required
* Would require additional glue code, because the second map is merged into the first